File: Lectures/Lecture 3/python/stack_list.py
class Stack:
    def __init__(self):
        self.stack = []

    def __str__(self):
        return str(self.stack)

    def push(self, item):
        self.stack.append(item)

    def pop(self):
        value = self.stack[-1]
        del self.stack[-1]
        return value

    def is_empty(self):
        return len(self.stack) == 0

    def peek(self):
        if len(self.stack) == 0:
            return None

        return self.stack[-1]

    def size(self):
        return len(self.stack)


if __name__ == '__main__':
    my_stack = Stack()
    print("my_stack:", my_stack)

    print("Empty?", my_stack.is_empty())

    print()

    my_stack.push("hello")
    print("my_stack:", my_stack)
    my_stack.push("world")
    print("my_stack:", my_stack)
    print("Empty?", my_stack.is_empty())

    print("peek:", my_stack.peek())
    print()

    print("pop:", my_stack.pop())
    print("my_stack:", my_stack)
    print("pop:", my_stack.pop())
    print("my_stack:", my_stack)
    print("peek:", my_stack.peek())

File: Lectures/Lecture 3/cpp/stack_list.cpp
#include <iostream>
#include <vector>

using namespace std;

class Stack
{
private:
    vector<string> stack;

public:
    Stack() {
    }

    friend ostream & operator << (ostream &out, const Stack &stack);
    void push(string text);
    string pop();
    string peek();
    bool is_empty() const {
        return (stack.size() == 0);
    }
    int size() {
        return stack.size();
    }
};

ostream & operator << (ostream &out, const Stack &stack) 
{
    out << "[";

    string separator;
    for (auto i: stack.stack) {
        out << separator << "'" << i << "'";
        separator = ", ";
    }
    out << "]";
    return out;
}

void Stack::push(string text) {
    this->stack.push_back(text);
}

string Stack::pop() {
    string text = this->stack.back();
    this->stack.pop_back();
    return text;
}

string Stack::peek() {
    if (this->size()==0) {
        // Todo: Better way of handling empty stack?
        return "";
    }
    return this->stack.back();
}

main () {
    Stack my_stack = Stack();
    cout << "My stack: " << my_stack << endl;

    cout << boolalpha;   // Ensure we can print booleans as true/false instead of 1/0
    cout << "Empty?: " << my_stack.is_empty() << endl;
    cout << endl;

    my_stack.push("hello");
    cout << "My stack: " << my_stack << endl;
    my_stack.push("world");
    cout << "My stack: " << my_stack << endl;
    cout << "Empty?: " << my_stack.is_empty() << endl;

    cout << "Peek: " << my_stack.peek() << endl;
    cout << endl;

    cout << "Pop: " << my_stack.pop() << endl;
    cout << "My stack: " << my_stack << endl;
    cout << "Pop: " << my_stack.pop() << endl;
    cout << "My stack: " << my_stack << endl;
    cout << "Peek:" << my_stack.peek() << endl;
}